libs/frontend/feature-build/src/lib/dnd/builder-dnd-component.ts

```diff
@@ -167,30 +179,57 @@ export const BuilderDndComponent = defineComponent({
         dndState.value,
       )
       const { editView, selectedComponent } = site.value.editor ?? {}
-      // Don't drag/drop root component, and disable when ProseMirror is active
-      const dndProps =
-        component.value.parent && !editView?.hasFocus()
-          ? {
-              draggable: !editView || selectedComponent?.id !== component.value.id,
-              ...dndState,
-              onDragstart: dnd.dragstart,
-              onDrag: dnd.drag,
-              onDragenter: dnd.dragenter,
-              onDragover: dnd.dragover,
-              onDragleave: dnd.dragleave,
-              onDrop: dnd.drop,
-              onDragend: dnd.dragend,
-            }
-          : {
-              onDragend: dnd.dragend,
-            }
+      const isRoot = !component.value.parent
+      const editViewFocus = editView?.hasFocus()
+      const droppableDndProps = {
+        onDragenter: dnd.dragenter,
+        onDragover: dnd.dragover,
+        onDragleave: dnd.dragleave,
+        onDrop: dnd.drop,
+        onDragend: dnd.dragend,
+      }
+
+      // If ProseMirror editor is active, only allow dragend event to be fired.
+      // If the component is root component, only allow hover&drop events to be fired.
+      // Otherwise allow all drag, hover, and drop events to be fired.
+      let dndProps: Record<string, unknown> = {}
+      if (editViewFocus) {
+        dndProps = { onDragend: dnd.dragend }
+      } else if (isRoot) {
+        dndProps = droppableDndProps
+      } else {
+        dndProps = {
+          draggable: !editView || selectedComponent?.id !== component.value.id,
+          onDragstart: dnd.dragstart,
+          onDrag: dnd.drag,
+          ...droppableDndProps,
+        }
+      }
+
+      if (
+        dndState.value?.hoverSelf ||
+        dndState.value?.hoverTop ||
+        dndState.value?.hoverRight ||
+        dndState.value?.hoverBottom ||
+        dndState.value?.hoverLeft
+      ) {
+        runtimeContext.buildDndState.value = {
+          hoverSelf: dndState.value.hoverSelf,
+          hoverTop: dndState.value.hoverTop,
+          hoverRight: dndState.value.hoverRight,
+          hoverBottom: dndState.value.hoverBottom,
+          hoverLeft: dndState.value.hoverLeft,
+          hoverCmpId: component.value.id,
+          hoverCmpParentIsRow: dndState.value.parentIsRow,
+        }
+      }
```

Comment:
We only want to show the overlay when a component is hovered in the editor rather than in the component tree, so I add a state to keep track of the D&D state in runtime context, and only update it in `builder-dnd-component`.